File: android/src/main/java/com/couchbase/lite/RNReplicatorConfiguration.kt
package com.couchbase.lite

import arrow.core.*
import arrow.core.extensions.either.applicative.applicative
import com.reactnativecouchbaselite.RNDatabase
import java.net.URI

object RNReplicatorConfiguration {
  fun decode(obj: SafeReadableMap): Either<String, ReplicatorConfiguration> =
    Either.applicative<String>().tupled(
      obj.getString("database"),
      obj.getString("target"),
      obj.getString("replicatorType").flatMap { TReplicatorType.decode(it) },
      obj.getMap("authenticator").fold(
        { Either.right(null) },
        { RNAuthenticator.decode(it) }
      ),
      obj.getArray("channels").fold(
        { Either.right(null) },
        { it.toListString() }
      )
    ).fix().map { tuple ->
      val database = RNDatabase.get(tuple.a)
      val endpoint = URLEndpoint(URI(tuple.b))
      val replicatorType = tuple.c
      val authenticator = tuple.d
      val channels = tuple.e

      val config = ReplicatorConfiguration(database, endpoint)

      config.replicatorType = replicatorType
      config.isContinuous = obj.getBoolean("continuous")

      if (channels != null) {
        config.channels = channels
      }

      if (authenticator != null) {
        config.authenticator = authenticator
      }
      config
    }
  }

File: ios/RNReplicatorConfiguration.swift
import Bow
import CouchbaseLiteSwift

class RNReplicatorConfiguration {
  static func decode(_ obj: SafeDictionary) -> Either<String, ReplicatorConfiguration> {
    return Either<String, ReplicatorConfiguration>.map(
      obj.getString("database").map(RNDatabase.get),
      obj.getString("target")
        .flatMap({ URL(string: $0).rightIfNotNull("target URL invalid") })
        .map({ URLEndpoint(url: $0) }),
      obj.getString("replicatorType")
        .flatMap(RNReplicatorType.decode)^,
      obj.getDict("authenticator").fold(
        { _ in Either.right(Option.none()) },
        { RNAuthenticator.decode($0).map(Option.some) }
      ),
      { (database, target, replicatorType, authenticator) in
        let config = ReplicatorConfiguration(database: database, target: target)
        
        config.replicatorType = replicatorType
        config.continuous = obj.getBool("continuous")
        
        obj.getArrayString("channels").map({
          config.channels = $0
        })
        
        authenticator.map({
          config.authenticator = $0
        })
        return config
      }
    )^
  }
}
